aiida/tools/graph/age_rules.py

```diff
@@ -0,0 +1,406 @@
+        # It starts empty.
+        new_results = operational_set.get_template()
+        self._iterations_done = 0
+        while (operational_set and self._iterations_done < self._maxiter):
+            self._iterations_done += 1
+            # loading results into new_results set:
+            self._load_results(new_results, operational_set)
+            # I set the active walkers to all results that have not been visited yet.
+            operational_set = new_results  #- self._accumulator_set
+        # I return a copy of the accumulator set, to ensure that changes made after do not
+        # affect this attribute of self:
+        return operational_set.copy()
+
+
+class RuleSaveWalkers(Operation):
+    """
+    Rule Save Walkers: when initialized, this rule will save a pointer
+    to an external stash variable. When run, this stash will be emptied
+    and a given operational_set will be saved there instead.
+    """
+
+    def __init__(self, stash):
+        self._stash = stash
+        super().__init__(max_iterations=1, track_edges=True)
+
+    def run(self, operational_set):
+        self._stash.empty()
+        self._stash += operational_set
+        return operational_set
+
+
+class RuleSetWalkers(Operation):
+    """
+    Rule Set Walkers: when initialized, this rule will save a pointer
+    to an external stash variable. When run, the given operational_set
+    will be emptied and the stash will be loaded in it.
+    """
+
+    def __init__(self, stash):
+        self._stash = stash
+        super().__init__(max_iterations=1, track_edges=True)
+
+    def run(self, operational_set):
+        operational_set.empty()
+        operational_set += self._stash
+        return operational_set
+
+
+class RuleSequence(Operation):
+    """
+    Rule Sequence is used to concatenate a series of rules together.
+    Concatenating querybuilders in a single rule its not enough because
+    one might want to stash results to perform two independent operations
+    in the starting set instead of a second operation from the results of
+    the first (see RuleSetWalkers and RuleSaveWalkers).
+    """
+
+    def __init__(self, rules, max_iterations=1):
+        for rule in rules:
+            if not isinstance(rule, Operation):
+                raise TypeError('rule has to be an instance of Operation-subclass')
+        self._rules = rules
+        self._accumulator_set = None
+        self._visits_set = None
+        super().__init__(max_iterations, track_edges=False)
+
+    def set_accumulator(self, accumulator_set):
+        self._accumulator_set = accumulator_set
+
+    def empty_accumulator(self):
+        if self._accumulator_set is not None:
+            self._accumulator_set.empty()
+
+    def set_visits(self, visits_set):
+        self._visits_set = visits_set
+
+    def empty_visits(self):
+        if self._visits_set is not None:
+            self._visits_set.empty()
+
+    def run(self, operational_set):
+        """run method"""
```

Comment:
pointless docstring :)

Reply:
So, question: we have `UpdateRule` and `ReplaceRule` that are subclases of `QueryRule`, and the difference is precisely (and almost exclusively) in the run method (one will accumulate visited nodes, the other will only returned the nodes visited in the last step). How would the docstrings of these three classes/methods work?

(1) If I want not to repeat describing the inputs, the docstring with the param description should be in the `QueryRule.run` (do the the other rules 'inherit' the docstring?), but if I want to describe the difference in the output then I should do particular docstrings for each `UpdateRule.run` and `ReplaceRule.run`.

(2) If I already mention the difference in `UpdateRule.run` and `ReplaceRule.run`, do I repeat this explanation in the docstrings of the classes?

Reply:
Ok, what I ended up doing is giving a general description on `QueryRule` and a more specific one on `ReplaceRule` and `UpdateRule`. For the method, I just described the abstract run method and the implementations inherited the docstring. I would have liked to do the same for the class docstrings but from what I could check they don't work the same way (i.e. `ReplaceRule` and `UpdateRule` won't inherit the `QueryRule` docstring).